src/Microsoft.CodeAnalysis.BannedApiAnalyzers/Core/RestrictedInternalsVisibleToAnalyzer.cs

```diff
@@ -0,0 +1,237 @@
+                    }
+
+                    VerifySymbol(symbol, context.Operation.Syntax, context.ReportDiagnostic);
+                },
+                OperationKind.ObjectCreation,
+                OperationKind.Invocation,
+                OperationKind.EventReference,
+                OperationKind.FieldReference,
+                OperationKind.MethodReference,
+                OperationKind.PropertyReference);
+
+            return;
+
+            // Local functions.
+            static ImmutableDictionary<IAssemblySymbol, ImmutableSortedSet<string>> GetRestrictedInternalsVisibleToMap(Compilation compilation)
+            {
+                var restrictedInternalsVisibleToAttribute = compilation.GetTypeByMetadataName("System.Runtime.CompilerServices.RestrictedInternalsVisibleToAttribute");
+                if (restrictedInternalsVisibleToAttribute == null)
+                {
+                    return ImmutableDictionary<IAssemblySymbol, ImmutableSortedSet<string>>.Empty;
+                }
+
+                var builder = ImmutableDictionary.CreateBuilder<IAssemblySymbol, ImmutableSortedSet<string>>();
+                foreach (var referencedAssemblySymbol in compilation.References.Select(compilation.GetAssemblyOrModuleSymbol).OfType<IAssemblySymbol>())
+                {
+                    // Check IVT
+                    if (!referencedAssemblySymbol.GivesAccessTo(compilation.Assembly))
+                    {
+                        continue;
+                    }
+
+                    var namespaceNameComparer = compilation.IsCaseSensitive ? StringComparer.Ordinal : StringComparer.OrdinalIgnoreCase;
+                    var namespaceBuilder = ImmutableSortedSet.CreateBuilder(namespaceNameComparer);
+                    foreach (var assemblyAttribute in referencedAssemblySymbol.GetAttributes())
+                    {
+                        // Look for ctor: "RestrictedInternalsVisibleToAttribute(string assemblyName, params string[] namespaces)"
+                        if (!Equals(assemblyAttribute.AttributeClass, restrictedInternalsVisibleToAttribute) ||
+                            assemblyAttribute.AttributeConstructor.Parameters.Length != 2 ||
+                            assemblyAttribute.AttributeConstructor.Parameters[0].Type.SpecialType != SpecialType.System_String ||
+                            !(assemblyAttribute.AttributeConstructor.Parameters[1].Type is IArrayTypeSymbol arrayType) ||
+                            arrayType.Rank != 1 ||
+                            arrayType.ElementType.SpecialType != SpecialType.System_String ||
+                            !assemblyAttribute.AttributeConstructor.Parameters[1].IsParams)
+                        {
+                            continue;
+                        }
+
+                        // Ensure the Restricted IVT is for the current compilation's assembly.
```

Comment:
Just curious, do we still need to do those checks if the expected attribute constructor is used?

Reply:
Just being defensive :)